File: py-sdk/mylogger/core/record.py
from dataclasses import dataclass, replace, field
from datetime import datetime
from enum import IntEnum
from typing import Optional, Dict, Any, List, Protocol
import uuid

class Level(IntEnum):
    DEBUG = 10
    INFO = 20
    WARN = 30
    ERROR = 40
    FATAL = 50

class Layer(IntEnum):
    PRESENTATION = 1
    GATEWAY = 2
    APPLICATION = 3
    DOMAIN = 4
    PERSISTENCE = 5
    INFRA = 6
    SECURITY = 7
    OBSERVABILITY = 8

@dataclass(frozen=True)
class LogRecord:
    message: str
    level: Level
    layer: Layer
    timestamp: datetime = field(default_factory=datetime.utcnow)
    record_id: str = field(default_factory=lambda: str(uuid.uuid4()))
    trace_id: Optional[str] = None
    span_id: Optional[str] = None
    service: Optional[str] = None
    env: Optional[str] = None
    context: Dict[str, Any] = field(default_factory=dict)

    def enrich(self, **kwargs) -> 'LogRecord':
        return replace(self, **kwargs)

    def to_dict(self) -> Dict[str, Any]:
        d = {k: v for k, v in self.__dict__.items()}
        d['level'] = self.level.name
        d['layer'] = self.layer.name
        d['timestamp'] = self.timestamp.isoformat()
        return d

File: ts-sdk/src/core/record.ts
import { Level, Layer } from './enums';

export interface LogRecord {
  readonly message: string;
  readonly level: Level;
  readonly layer: Layer;
  readonly timestamp: string;
  readonly record_id: string;
  readonly trace_id: string | null;
  readonly span_id: string | null;
  readonly service: string | null;
  readonly env: string | null;
  readonly context: Record<string, any>;
}

export abstract class BaseRecord implements LogRecord {
  readonly message: string;
  readonly level: Level;
  readonly layer: Layer;
  readonly timestamp: string;
  readonly record_id: string;
  readonly trace_id: string | null;
  readonly span_id: string | null;
  readonly service: string | null;
  readonly env: string | null;
  readonly context: Record<string, any>;

  constructor(data: LogRecord) {
    this.message = data.message;
    this.level = data.level;
    this.layer = data.layer;
    this.timestamp = data.timestamp;
    this.record_id = data.record_id;
    this.trace_id = data.trace_id;
    this.span_id = data.span_id;
    this.service = data.service;
    this.env = data.env;
    this.context = { ...data.context };
    Object.freeze(this);
    Object.freeze(this.context);
  }

  abstract toDict(): Record<string, any>;
  
  enrich(update: Partial<LogRecord>): this {
    const Constructor = this.constructor as any;
    return new Constructor({ ...this, ...update });
  }
}
